imports/wasi_snapshot_preview1/fs.go

```diff
@@ -1397,12 +1508,73 @@ func openFlags(oflags, fdflags uint16) (openFlags int, isDir bool) {
 // contents of a symbolic link.
 //
 // See: https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-path_readlinkfd-fd-path-string-buf-pointeru8-buf_len-size---errno-size
-var pathReadlink = stubFunction(
-	PathReadlinkName,
+var pathReadlink = newHostFunc(
+	PathReadlinkName, pathReadlinkFn,
 	[]wasm.ValueType{i32, i32, i32, i32, i32, i32},
 	"fd", "path", "path_len", "buf", "buf_len", "result.bufused",
 )
 
+func pathReadlinkFn(_ context.Context, mod api.Module, params []uint64) Errno {
+	fsc := mod.(*wasm.CallContext).Sys.FS()
+
+	fd := uint32(params[0])
+	path := uint32(params[1])
+	pathLen := uint32(params[2])
+	bufPtr := uint32(params[3])
+	bufLen := uint32(params[4])
+	resultBufUsedPtr := uint32(params[5])
+
+	if pathLen == 0 || bufLen == 0 {
+		return ErrnoInval
+	}
+
+	mem := mod.Memory()
+	buf, ok := mem.Read(bufPtr, bufLen)
+	if !ok {
+		return ErrnoFault
+	}
+
+	pathNameBuf, ok := mem.Read(path, pathLen)
+	if !ok {
+		return ErrnoFault
+	}
+
+	pathName := bufToStr(pathNameBuf, int(pathLen))
+
+	dir, ok := fsc.LookupFile(fd)
+	if !ok {
+		return ErrnoBadf // closed
+	} else if !dir.IsDir() {
+		return ErrnoNotdir
+	}
+
+	n, err := fsc.FS().Readlink(pathutil.Join(dir.Name, pathName), buf)
+	if err != nil {
+		return ToErrno(err)
+	}
+
+	// pathReadlinkFn requires the result to be the relative to this directory pointed by the file descriptor,
+	// so trim it.
+	// TODO: come up the better abstraction so that we could this trimming all at once in Readlink call ^.
+	if dirLen := len(dir.Name); dirLen > 0 {
```

Comment:
this seems like it is handled at the wrong abstraction. why is this happening here instead of inside syfs?